src/probability/martingale/borel_cantelli.lean

```diff
@@ -0,0 +1,491 @@
+      ⟨n, ⟨zero_le _, le_rfl⟩, h₂⟩ }
+end
+
+lemma least_ge_succ_eq_iff' (n : ℕ) {r : ℝ} :
+  least_ge f r (n + 1) ω = n + 1 ↔ least_ge f r n ω = n ∧ f n ω < r :=
+begin
+  split,
+  { intro h,
+    have : least_ge f r n ω = n,
+    { refine le_antisymm (hitting_le _) _,
+      by_contra hlt,
+      rw [not_le, least_ge] at hlt,
+      refine ne_of_lt _ h,
+      rw [least_ge, hitting_lt_iff (n + 1) le_rfl],
+      exact ⟨least_ge f r n ω, ⟨zero_le _, nat.lt_succ_of_le (hitting_le _)⟩,
+        hitting_mem_set_of_hitting_lt hlt⟩ },
+    refine ⟨this, _⟩,
+    by_contra h',
+    rw not_lt at h',
+    rw ((least_ge_succ_eq_iff n).2 ⟨this, h'⟩) at h,
+    norm_num at h },
+  { rintro ⟨h₁, h₂⟩,
+    refine le_antisymm (hitting_le _) (nat.succ_le_of_lt _),
+    by_contra h,
+    have : least_ge f r (n + 1) ω = least_ge f r n ω :=
+      le_antisymm (h₁.symm ▸ not_lt.1 h) (hitting_mono n.le_succ),
+    rw h₁ at this,
+    refine not_lt.2 _ h₂,
+    refine this ▸ hitting_mem_set_of_hitting_lt _,
+    rw [← least_ge, this],
+    exact n.lt_succ_self },
+end
+
+-- The following four lemmas shows `least_ge` behaves like a stopped process. Ideally we should
+-- define `least_ge` as a stopping time and take its stopped process. However, we can't do that
+-- with our current definition since a stopping time takes only finite indicies. An upcomming
+-- refactor should hopefully make it possible to have stopping times taking infinity as a value
+lemma least_ge_mono {n m : ℕ} (hnm : n ≤ m) (r : ℝ) (ω : Ω) :
+  least_ge f r n ω ≤ least_ge f r m ω :=
+hitting_mono hnm
+
+lemma least_ge_eq_min (π : Ω → ℕ) (r : ℝ) (ω : Ω)
+  {n : ℕ} (hπn : ∀ ω, π ω ≤ n) :
+  least_ge f r (π ω) ω = min (π ω) (least_ge f r n ω) :=
+begin
+  classical,
+  refine le_antisymm (le_min (least_ge_le _) (least_ge_mono (hπn ω) r ω)) _,
+  by_cases hle : π ω ≤ least_ge f r n ω,
+  { rw [min_eq_left hle, least_ge],
+    by_cases h : ∃ j ∈ set.Icc 0 (π ω), f j ω ∈ set.Ici r,
+    { refine hle.trans (eq.le _),
+      rw [least_ge, ← hitting_eq_hitting_of_exists (hπn ω) h] },
+    { simp only [hitting, if_neg h] } },
+  { rw [min_eq_right (not_le.1 hle).le, least_ge, least_ge,
+      ← hitting_eq_hitting_of_exists (hπn ω) _],
+    rw [not_le, least_ge, hitting_lt_iff _ (hπn ω)] at hle,
+    exact let ⟨j, hj₁, hj₂⟩ := hle in ⟨j, ⟨hj₁.1, hj₁.2.le⟩, hj₂⟩ }
+end
+
+lemma least_ge_apply_eq_stopped_value (f : ℕ → Ω → ℝ) (π : Ω → ℕ) (r : ℝ) (ω : Ω)
+  {n : ℕ} (hπn : ∀ ω, π ω ≤ n) :
+  f (least_ge f r (π ω) ω) ω = stopped_value f (λ ω, min (π ω) (least_ge f r n ω)) ω :=
+by { rw least_ge_eq_min _ _ _ hπn, simp_rw [stopped_value], }
+
+lemma stopped_value_stopped_value_least_ge (f : ℕ → Ω → ℝ) (π : Ω → ℕ) (r : ℝ)
+  {n : ℕ} (hπn : ∀ ω, π ω ≤ n) :
+  stopped_value (λ i, stopped_value f (least_ge f r i)) π
+    = (λ ω, stopped_value f (λ ω, min (π ω) (least_ge f r n ω)) ω) :=
+begin
+  ext1 ω,
+  exact least_ge_apply_eq_stopped_value f π r ω hπn,
+end
```

Comment:
```suggestion
lemma stopped_value_stopped_value_least_ge (f : ℕ → Ω → ℝ) (π : Ω → ℕ) (r : ℝ)
  {n : ℕ} (hπn : ∀ ω, π ω ≤ n) :
  stopped_value (λ i, stopped_value f (least_ge f r i)) π
    = stopped_value (stopped_process f (least_ge f r n)) π :=
by { ext1 ω, simp_rw [stopped_process, stopped_value], rw least_ge_eq_min _ _ _ hπn, }
```
This formulation makes clear that `least_ge` behaves like a stopped process. You don't need to change anything in other proofs because the new way to write the result is defeq to the old one.